lib/galaxy/security/passwords.py

```diff
@@ -57,39 +51,19 @@ def check_password_PBKDF2(guess, hashed):
     # Split the database representation to extract cost_factor and salt
     name, hash_function, cost_factor, salt, encoded_original = hashed.split('$', 5)
     # Hash the guess using the same parameters
-    hashed_guess = pbkdf2_bin(guess, salt, int(cost_factor), KEY_LENGTH, getattr(hashlib, hash_function))
+    hashed_guess = pbkdf2_bin(guess, salt, int(cost_factor), KEY_LENGTH, hash_function)
     encoded_guess = unicodify(b64encode(hashed_guess))
     return safe_str_cmp(encoded_original, encoded_guess)
 
 
-# Taken from https://github.com/mitsuhiko/python-pbkdf2/blob/master/pbkdf2.py
-# (c) Copyright 2011 by Armin Ronacher, BSD LICENSE
-_pack_int = Struct('>I').pack
-
-
-def pbkdf2_bin(data, salt, iterations=1000, keylen=24, hashfunc=None):
+def pbkdf2_bin(data, salt, iterations=COST_FACTOR, keylen=24, hashfunc=None):
     """Returns a binary digest for the PBKDF2 hash algorithm of `data`
     with the given `salt`.  It iterates `iterations` time and produces a
-    key of `keylen` bytes.  By default SHA-1 is used as hash function,
+    key of `keylen` bytes.  By default SHA-256 is used as hash function,
     a different hashlib `hashfunc` can be provided.
     """
-    hashfunc = hashfunc or hashlib.sha1
-    mac = hmac.new(smart_str(data), None, hashfunc)
-
-    def _pseudorandom(x, mac=mac):
-        h = mac.copy()
-        h.update(x)
-        digest = h.digest()
-        if six.PY2:
-            return digest, [ord(_) for _ in digest]
-        return digest, digest
+    hashfunc = hashfunc or HASH_FUNCTION
+    data = bytes(smart_str(data))
```

Comment:
`smart_str()` should already return a bytestring.

```suggestion
    data = smart_str(data)
```

Reply:
D'oh, thanks.  I swapped in our `smart_str` here when adapting the werkzeug code I was looking at and forgot to drop `bytes`.